src/python/pants/goal/migrate_call_by_name.py

```diff
@@ -3,35 +3,655 @@
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>) -> the_rule_to_call(**implicitly())"""
+
+        logger.debug(ast.dump(get, indent=2))
+        assert len(get.args) == 1, f"Expected 1 arg, got {len(get.args)}"
+        assert isinstance(output_type := get.args[0], ast.Name), f"Expected Name, got {get.args[0]}"
+
+        dep = next(
+            dep
+            for dep in deps
+            if dep["output_type"]["name"] == output_type.id and not dep["input_types"]
+        )
+        module = dep["rule_dep"]["module"]
+        new_function = dep["rule_dep"]["function"]
+
+        new_call = ast.Call(
+            func=ast.Name(id=new_function),
+            args=[],
+            keywords=[
+                ast.keyword(value=ast.Call(func=ast.Name(id="implicitly"), args=[], keywords=[]))
+            ],
+        )
+        imports = [ast.ImportFrom(module, names=[ast.alias(new_function)], level=0)]
+        return new_call, imports
+
+    def map_long_form_get_to_new_syntax(
+        self, get: ast.Call, deps: list[RuleGraphGetDep]
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>, <InputType>, input) -> the_rule_to_call(**implicitly(input))"""
+
+        logger.debug(ast.dump(get, indent=2))
+        assert len(get.args) == 3, f"Expected 3 arg, got {len(get.args)}"
+        assert isinstance(output_type := get.args[0], ast.Name), f"Expected Name, got {get.args[0]}"
+        assert isinstance(input_type := get.args[1], ast.Name), f"Expected Name, got {get.args[1]}"
+
+        dep = next(
+            dep
+            for dep in deps
+            if dep["output_type"]["name"] == output_type.id
+            and len(dep["input_types"]) == 1
+            and dep["input_types"][0]["name"] == input_type.id
+        )
+        module = dep["rule_dep"]["module"]
+        new_function = dep["rule_dep"]["function"]
+
+        new_call = ast.Call(
+            func=ast.Name(id=new_function),
+            args=[],
+            keywords=[
+                ast.keyword(
+                    value=ast.Call(
+                        func=ast.Name(id="implicitly"),
+                        args=[ast.Dict(keys=[get.args[2]], values=[ast.Name(id=input_type.id)])],
+                        keywords=[],
+                    )
+                )
+            ],
+        )
+        imports = [ast.ImportFrom(module, names=[ast.alias(new_function)], level=0)]
+        return new_call, imports
+
+    def map_short_form_get_to_new_syntax(
+        self, get: ast.Call, deps: list[RuleGraphGetDep]
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>, <InputType>(<constructor args for input>)) -> the_rule_to_call(input,
+
+        **implicitly())
+        """
+
+        logger.debug(ast.dump(get, indent=2))
+        assert len(get.args) == 2, f"Expected 2 arg, got {len(get.args)}"
+        assert isinstance(output_type := get.args[0], ast.Name), f"Expected Name, got {get.args[0]}"
```

Comment:
This is a neat pattern!

Reply:
walrus FTW

Though, the lack of AST inference is brutal - I need to extract some utility AST stuff to support our use cases - too much dupe code to make the type checker happy (but, the value/need in making the type checker happy with AST code is really high)